File: src/main/java/com/soapdogg/lispInterpreter/singleton/EvaluatorSingleton.java
package com.soapdogg.lispInterpreter.singleton;

import com.soapdogg.lispInterpreter.evaluator.AtomNodeEvaluator;
import com.soapdogg.lispInterpreter.evaluator.CondFunctionEvaluator;
import com.soapdogg.lispInterpreter.evaluator.NodeEvaluator;
import com.soapdogg.lispInterpreter.evaluator.ProgramEvaluator;
import lombok.Getter;

@Getter
public enum EvaluatorSingleton {
    INSTANCE;

    private final AtomNodeEvaluator atomNodeEvaluator;
    private final NodeEvaluator nodeEvaluator;
    private final CondFunctionEvaluator condFunctionEvaluator;
    private final ProgramEvaluator programEvaluator;

    EvaluatorSingleton() {
        atomNodeEvaluator = AtomNodeEvaluator.newInstance();
        nodeEvaluator = NodeEvaluator.newInstance(
            atomNodeEvaluator,
            DeterminerSingleton.INSTANCE.getExpressionNodeDeterminer(),
            DeterminerSingleton.INSTANCE.getUserDefinedFunctionNameDeterminer(),
            AsserterSingleton.INSTANCE.getFunctionLengthAsserter()
        );
        condFunctionEvaluator = CondFunctionEvaluator.newInstance(
            ValueRetrieverSingleton.INSTANCE.getListValueRetriever(),
            nodeEvaluator,
            ComparatorSingleton.INSTANCE.getNodeValueComparator()
        );
        programEvaluator = ProgramEvaluator.newInstance(
            DeterminerSingleton.INSTANCE.getExpressionNodeDeterminer(),
            AsserterSingleton.INSTANCE.getAtomRootNodeAsserter(),
            nodeEvaluator
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/singleton/EvaluatorSingleton.kt
package com.soapdogg.lispInterpreter.singleton

import com.soapdogg.lispInterpreter.evaluator.*

enum class EvaluatorSingleton {
    INSTANCE;

    private val topProgramStackItemCreator = TopProgramStackItemCreator(
        GeneratorSingleton.INSTANCE.programStackItemGenerator
    )

    private val topProgramStackItemUpdater = TopProgramStackItemUpdater(
        GeneratorSingleton.INSTANCE.programStackItemGenerator
    )

    private val postEvaluationStackUpdater = PostEvaluationStackUpdater(
        topProgramStackItemUpdater
    )

    private val builtInFunctionEvaluator = BuiltInFunctionEvaluator(
        FunctionSingleton.INSTANCE.functionMap,
        postEvaluationStackUpdater
    )

    private val condChildStackItemBuilder = CondChildStackItemBuilder(
        GeneratorSingleton.INSTANCE.nodeGenerator,
        topProgramStackItemCreator
    )

    private val condProgramStackItemEvaluator = CondFunctionEvaluator(
        topProgramStackItemUpdater,
        condChildStackItemBuilder
    )

    private val quoteFunctionEvaluator = QuoteFunctionEvaluator(
        postEvaluationStackUpdater
    )

    private val stackUpdateDeterminer = StackUpdateDeterminer(
        topProgramStackItemCreator,
        postEvaluationStackUpdater
    )

    private val condChildFunctionEvaluator = CondChildFunctionEvaluator(
        stackUpdateDeterminer
    )

    private val userDefinedFunctionEvaluator = UserDefinedFunctionEvaluator(
        stackUpdateDeterminer
    )

    private val unfinishedProgramStackItemEvaluator = UnfinishedProgramStackItemEvaluator(
        stackUpdateDeterminer
    )

    private val finishedProgramStackItemEvaluator = FinishedProgramStackItemEvaluator(
        postEvaluationStackUpdater,
        FunctionSingleton.INSTANCE.functionMap,
        builtInFunctionEvaluator,
        userDefinedFunctionEvaluator
    )

    private val nodeEvaluatorIterative = RootNodeEvaluator(
        topProgramStackItemCreator,
        condProgramStackItemEvaluator,
        condChildFunctionEvaluator,
        quoteFunctionEvaluator,
        unfinishedProgramStackItemEvaluator,
        finishedProgramStackItemEvaluator
    )

    val programEvaluator: ProgramEvaluator = ProgramEvaluator(
        AsserterSingleton.INSTANCE.atomRootNodeAsserter,
        nodeEvaluatorIterative
    )
}